pkg/providers/ovirt/client/ovirt-client.go

```diff
@@ -0,0 +1,272 @@
+type OvirtClient interface {
+	GetVM(id *string, name *string, cluster *string) (*ovirtsdk.Vm, error)
+	Close() error
+}
+
+// ConnectionSettings wrap information required to make oVirt API connection
+type ConnectionSettings struct {
+	URL      string
+	Username string
+	Password string
+	Insecure bool
+	CAFile   string
+}
+
+// RichOvirtClient is responsilbe for retrieving VM data from oVirt API
+type richOvirtClient struct {
+	connection *ovirtsdk.Connection
+}
+
+// NewRitchOvirtClient creates new, connected rich oVirt client. After it is no longer needed, call Close().
+func NewRitchOvirtClient(cs *ConnectionSettings) (OvirtClient, error) {
+	con, err := connect(cs.URL, cs.Username, cs.Password, cs.CAFile, cs.Insecure)
+	if err != nil {
+		return nil, err
+	}
+	client := richOvirtClient{
+		connection: con,
+	}
+	return &client, nil
+}
+
+// Close releases the resources used by this client.
+func (client *richOvirtClient) Close() error {
+	return client.connection.Close()
+}
+
+// GetVM rertrieves oVirt VM data for given id or name and cluster. VM will have certain links followed and updated.
+func (client *richOvirtClient) GetVM(id *string, name *string, cluster *string) (*ovirtsdk.Vm, error) {
+	vm, err := client.fetchVM(id, name, cluster)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateNics(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateDiskAttachments(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateCdRoms(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateFloppies(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateHostDevices(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateReportedDevices(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateQuota(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateWatchdogs(vm)
+	if err != nil {
+		return nil, err
+	}
+	err = client.populateGraphicsConsoles(vm)
+	if err != nil {
+		return nil, err
+	}
+	return vm, nil
+}
+
+func (client *richOvirtClient) fetchVM(id *string, name *string, cluster *string) (*ovirtsdk.Vm, error) {
+	// Id of the VM specified:
+	if id != nil {
+		response, err := client.connection.SystemService().VmsService().VmService(*id).Get().Send()
+		if err != nil {
+			return nil, err
+		}
+		if vm, ok := response.Vm(); ok {
+			return vm, nil
+		}
+		return nil, fmt.Errorf("Virtual machine %v not found", *id)
+	}
+	// Cluster/name of the VM specified:
+	response, err := client.connection.SystemService().VmsService().List().Search(fmt.Sprintf("name=%v and cluster=%v", *name, *cluster)).Send()
+	if err != nil {
+		return nil, err
+	}
+	vms, _ := response.Vms()
+	vmsCount := len(vms.Slice())
+	if vmsCount == 1 {
+		return vms.Slice()[0], nil
+	} else if vmsCount > 1 {
+		return nil, fmt.Errorf("Found more than one virtual machine %v in cluster %v. VM IDs: %v", *name, *cluster, getVMIDs(vms.Slice()))
```

Comment:
minor comment: the vms will probably be in different clusters share has the same name.
Perhaps: `Found more than one virtual machine with name %v in clusters with name %v`

Reply:
:+1: Fixed.